cmd/private-org-peribolos-sync/main.go

```diff
@@ -0,0 +1,233 @@
+	err := validateOptions(&o)
+	if err != nil {
+		logrus.WithError(err).Fatal("invalid options")
+	}
+	logger := logrus.WithField("destination-org", o.destOrg)
+
+	sigTerm := make(chan os.Signal, 1)
+	signal.Notify(sigTerm, syscall.SIGTERM)
+	go func() {
+		for {
+			select {
+			case <-sigTerm:
+				logger.Infof("Shutdown signal received, exiting...")
+				os.Exit(1)
+			}
+		}
+	}()
+
+	b, err := ioutil.ReadFile(o.peribolosConfig)
+	if err != nil {
+		logger.WithError(err).Fatal("could not read peribolos configuration file")
+	}
+
+	var peribolosConfig org.FullConfig
+	if err := yaml.Unmarshal(b, &peribolosConfig); err != nil {
+		logger.WithError(err).Fatal("failed to unmarshal peribolos config")
+	}
+
+	secretAgent := &secret.Agent{}
+	if err := secretAgent.Start([]string{o.github.TokenPath}); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
+	}
+	gc, err := o.github.GitHubClient(secretAgent, false)
+	if err != nil {
+		logger.WithError(err).Fatal("Error getting GitHub client.")
+	}
+
+	orgRepos, err := getOrgReposWithOfficialImages(o.releaseRepoPath, o.includeReposSet)
+	if err != nil {
+		logger.WithError(err).Fatal("couldn't get the list of org/repos that promote official images")
+	}
+
+	peribolosRepos := generateRepositories(gc, orgRepos, logger)
+	peribolosConfigByOrg := peribolosConfig.Orgs[o.destOrg]
+	peribolosConfigByOrg.Repos = peribolosRepos
+	peribolosConfig.Orgs[o.destOrg] = peribolosConfigByOrg
+
+	out, err := yaml.Marshal(peribolosConfig)
+	if err != nil {
+		logrus.WithError(err).Fatalf("%s failed to marshal output.", o.peribolosConfig)
+	}
+
+	if err := ioutil.WriteFile(o.peribolosConfig, out, 0666); err != nil {
+		logrus.WithError(err).Fatal("Failed to write output.")
+	}
+}
+
+func generateRepositories(gc gitHubClient, orgRepos map[string]sets.String, logger *logrus.Entry) map[string]org.Repo {
+	peribolosRepos := make(map[string]org.Repo)
+
+	for orgName, repos := range orgRepos {
+		for repo := range repos {
+			logger.WithFields(logrus.Fields{"org": orgName, "repo": repo}).Info("Processing repository details...")
+
+			fullRepo, err := gc.GetRepo(orgName, repo)
+			if err != nil {
+				logger.WithError(err).Fatal("couldn't get repo details")
+			}
+
+			peribolosRepos[fullRepo.Name] = org.PruneRepoDefaults(org.Repo{
+				Description:      &fullRepo.Description,
+				HomePage:         &fullRepo.Homepage,
+				Private:          &fullRepo.Private,
+				HasIssues:        &fullRepo.HasIssues,
+				HasProjects:      &fullRepo.HasProjects,
+				HasWiki:          &fullRepo.HasWiki,
+				AllowMergeCommit: &fullRepo.AllowMergeCommit,
+				AllowSquashMerge: &fullRepo.AllowSquashMerge,
+				AllowRebaseMerge: &fullRepo.AllowRebaseMerge,
+				Archived:         &fullRepo.Archived,
+				DefaultBranch:    &fullRepo.DefaultBranch,
+			})
+		}
+	}
+
+	return peribolosRepos
+}
+
+func getOrgReposWithOfficialImages(releaseRepoPath string, includeRepos map[string]sets.String) (map[string]sets.String, error) {
```

Comment:
The name of this method lies, it does more than that and should be actually called `getOrgReposWithOfficialImagesAndThoseInWhitelist`. That is of course a very bad name, which indicates the code might not be structured the best way.

I suggest to do one of the following:

1) **Either** keep the current structure but rename this to something more generic (`getReposForPrivateOrg`) and document it includes repos from two sources
2) **or** keep the name, remove the whitelist part from the list, and extend the returned map with the whitelisted repos at the callsite, like:

```
reposForPrivateOrg, err = getOrgReposWithOfficialImages(releaseRepoPath string);
...check errors...
reposForPrivateOrg = addWhitelist(reposForPrivateOrg, whitelist)
```

Reply:
Option 1 is the best in this case. If we go with option 2 we will end up relying on the user that the whitelisted repo actually exists and we will make unnecessary GitHub calls.
I would prefer to allow picking the repo while we iterate in the release repo.

Reply:
That sounds good!